internal/btf/btf.go

```diff
@@ -156,68 +156,32 @@ func loadSpecFromELF(file *internal.SafeELFFile, variableOffsets map[variable]ui
 		return spec, nil
 	}
 
-	if err := spec.loadExtInfos(btfExtSection); err != nil {
-		return nil, fmt.Errorf("can't parse ext info: %w", err)
-	}
-
-	return spec, nil
-}
-
-// loadExtInfos parses the .BTF.ext section and populates spec's funcInfo,
-// lineInfo and coreRelos fields using the contents.
-func (spec *Spec) loadExtInfos(sec *elf.Section) error {
-	// Open unbuffered section reader. binary.Read() calls io.ReadFull on
-	// the header structs, resulting in one syscall per header.
-	r := sec.Open()
-
-	extHeader, err := parseBTFExtHeader(r, spec.byteOrder)
-	if err != nil {
-		return fmt.Errorf("parsing BTF extension header: %w", err)
-	}
-
-	coreHeader, err := parseBTFExtCoreHeader(r, spec.byteOrder, extHeader)
-	if err != nil {
-		return fmt.Errorf("parsing BTF CO-RE header: %w", err)
+	if btfExtSection.ReaderAt == nil {
+		return nil, fmt.Errorf("compressed ext_info is not supported")
 	}
 
-	buf := internal.NewBufferedSectionReader(sec, extHeader.funcInfoStart(), int64(extHeader.FuncInfoLen))
-	funcInfos, err := parseFuncInfos(buf, spec.byteOrder, spec.strings)
+	extInfo, err := loadExtInfos(btfExtSection, file.ByteOrder, spec.strings)
 	if err != nil {
-		return fmt.Errorf("parsing BTF function info: %w", err)
-	}
-
-	buf = internal.NewBufferedSectionReader(sec, extHeader.lineInfoStart(), int64(extHeader.LineInfoLen))
-	lineInfos, err := parseLineInfos(buf, spec.byteOrder, spec.strings)
-	if err != nil {
-		return fmt.Errorf("parsing BTF line info: %w", err)
-	}
-
-	relos := make(map[string]CoreRelos)
-	if coreHeader != nil && coreHeader.CoreReloOff > 0 && coreHeader.CoreReloLen > 0 {
-		buf = internal.NewBufferedSectionReader(sec, extHeader.coreReloStart(coreHeader), int64(coreHeader.CoreReloLen))
-		relos, err = parseCoreRelos(buf, spec.byteOrder, spec.strings)
-		if err != nil {
-			return fmt.Errorf("parsing CO-RE relocation info: %w", err)
-		}
+		return nil, fmt.Errorf("can't parse ext info: %w", err)
 	}
 
-	if err := spec.splitExtInfos(funcInfos, lineInfos, relos); err != nil {
-		return fmt.Errorf("linking funcInfos and lineInfos: %w", err)
+	if err := spec.splitExtInfos(extInfo); err != nil {
+		return nil, fmt.Errorf("linking funcInfos and lineInfos: %w", err)
 	}
 
-	return nil
+	return spec, nil
 }
 
 // splitExtInfos takes FuncInfos, LineInfos and CoreRelos indexed by section and
 // transforms them to be indexed by function. Retrieves function names from
 // the BTF spec.
-func (spec *Spec) splitExtInfos(sfi map[string]FuncInfos, sli map[string]LineInfos, scr map[string]CoreRelos) error {
+func (spec *Spec) splitExtInfos(info *extInfo) error {
```

Comment:
`extInfo` is neat. :+1: I think we should probably rename this to something like `assignExtInfo`, though.

Reply:
I think `splitExtInfos` is better, since it implies that there is a transformation going on here.